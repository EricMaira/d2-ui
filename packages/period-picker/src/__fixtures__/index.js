export const periodTypes = {
    Daily: 'Daily',
    Weekly: 'Weekly',
    WeeklyWednesday: 'Weekly Wednesday',
    WeeklyThursday: 'Weekly Thursday',
    WeeklySaturday: 'Weekly Saturday',
    WeeklySunday: 'Weekly Sunday',
    BiWeekly: 'Bi weekly',
    Monthly: 'Monthly',
    BiMonthly: 'Bi-monthly',
    Quarterly: 'Quarterly',
    SixMonthly: 'Six monthly',
    SixMonthlyApril: 'Six monthly starting in April',
    SixMonthlyNov: 'Six monthly starting in November',
    Yearly: 'Yearly',
    FinancialApril: 'Financial year starting in April',
    FinancialJuly: 'Financial year starting in July',
    FinancialOct: 'Financial year starting in October',
    FinancialNov: 'Financial year starting in November',
};

export const getPeriodTypesResponse = {
    periodTypes: [
        {
            frequencyOrder: 1,
            name: 'Daily',
            isoDuration: 'P1D',
            isoFormat: 'yyyyMMdd',
        },
        {
            frequencyOrder: 7,
            name: 'Weekly',
            isoDuration: 'P7D',
            isoFormat: 'yyyyWn',
        },
        {
            frequencyOrder: 7,
            name: 'WeeklyWednesday',
            isoDuration: 'P7D',
            isoFormat: 'yyyyWedWn',
        },
        {
            frequencyOrder: 7,
            name: 'WeeklyThursday',
            isoDuration: 'P7D',
            isoFormat: 'yyyyThuWn',
        },
        {
            frequencyOrder: 7,
            name: 'WeeklySaturday',
            isoDuration: 'P7D',
            isoFormat: 'yyyySatWn',
        },
        {
            frequencyOrder: 7,
            name: 'WeeklySunday',
            isoDuration: 'P7D',
            isoFormat: 'yyyySunWn',
        },
        {
            frequencyOrder: 14,
            name: 'BiWeekly',
            isoDuration: 'P14D',
            isoFormat: 'yyyyBiWn',
        },
        {
            frequencyOrder: 30,
            name: 'Monthly',
            isoDuration: 'P1M',
            isoFormat: 'yyyyMM',
        },
        {
            frequencyOrder: 61,
            name: 'BiMonthly',
            isoDuration: 'P2M',
            isoFormat: 'yyyyMMB',
        },
        {
            frequencyOrder: 91,
            name: 'Quarterly',
            isoDuration: 'P3M',
            isoFormat: 'yyyyQn',
        },
        {
            frequencyOrder: 182,
            name: 'SixMonthly',
            isoDuration: 'P6M',
            isoFormat: 'yyyySn',
        },
        {
            frequencyOrder: 182,
            name: 'SixMonthlyApril',
            isoDuration: 'P6M',
            isoFormat: 'yyyyAprilSn',
        },
        {
            frequencyOrder: 182,
            name: 'SixMonthlyNov',
            isoDuration: 'P6M',
            isoFormat: 'yyyyNovSn',
        },
        {
            frequencyOrder: 365,
            name: 'Yearly',
            isoDuration: 'P1Y',
            isoFormat: 'yyyy',
        },
        {
            frequencyOrder: 365,
            name: 'FinancialApril',
            isoDuration: 'P1Y',
            isoFormat: 'yyyyApril',
        },
        {
            frequencyOrder: 365,
            name: 'FinancialJuly',
            isoDuration: 'P1Y',
            isoFormat: 'yyyyJuly',
        },
        {
            frequencyOrder: 365,
            name: 'FinancialOct',
            isoDuration: 'P1Y',
            isoFormat: 'yyyyOct',
        },
        {
            frequencyOrder: 365,
            name: 'FinancialNov',
            isoDuration: 'P1Y',
            isoFormat: 'yyyyNov',
        },
    ],
};
