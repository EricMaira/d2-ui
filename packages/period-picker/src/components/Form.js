import i18n from '@dhis2/d2-i18n';
import FormHelperText from '@material-ui/core/FormHelperText';
import { withStyles } from '@material-ui/core/styles';
import parsePeriod from 'd2/period/parser';
import PropTypes from 'prop-types';
import React, { Fragment } from 'react';
import PeriodFields from './PeriodFields';
import Select from './Select';

const styles = theme => {
    return {
        error: {
            color: theme.colors.negative,
        },
        helper: {
            margin: 8,
        },
    };
};

export const Form = ({
    periodTypes,
    periodType,
    onChange,
    getFieldValue,
    errorText,
    value,
    classes,
}) => (
    <Fragment>
        {periodTypes && (
            <Select
                name="periodType"
                label={i18n.t('Period type')}
                value={periodType}
                onChange={onChange}
                options={periodTypes}
            />
        )}
        {periodType && (
            <PeriodFields
                periodType={periodType}
                onChange={onChange}
                getValue={getFieldValue}
            />
        )}
        {errorText && (
            <FormHelperText className={`${classes.error} ${classes.helper}`}>
                {errorText}
            </FormHelperText>
        )}
        {value &&
            !errorText && (
                <FormHelperText className={classes.helper}>
                    {parsePeriod(value).name}
                </FormHelperText>
            )}
    </Fragment>
);

Form.propTypes = {
    periodTypes: PropTypes.object,
    periodType: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    getFieldValue: PropTypes.func.isRequired,
    errorText: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(Form);
