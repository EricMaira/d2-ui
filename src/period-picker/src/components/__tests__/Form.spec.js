import React from 'react';
import { shallow } from 'enzyme';
import { periodTypes } from '../../__fixtures__';
import { Form } from '../Form';

describe('<Form/>', () => {
    const defaultProps = {
        periodTypes,
        periodType: '',
        onChange: jest.fn(),
        getFieldValue: jest.fn(),
        errorText: '',
        value: '',
        classes: {
            error: 'error',
            helper: 'helper',
        },
    };
    it('Renders the only the periodType <Select/> when periodType is empty', () => {
        const wrapper = shallow(<Form {...defaultProps} />);
        expect(wrapper.find('Fragment').children().length).toEqual(1);
        expect(wrapper.find('WithStyles(Select)').length).toEqual(1);
    });

    it('Renders <PeriodFields/> below the <Select/> when a periodType is present', () => {
        const props = { ...defaultProps, periodType: 'Daily' };
        const wrapper = shallow(<Form {...props} />);
        expect(wrapper.find('Fragment').children().length).toEqual(2);
        expect(wrapper.find('WithStyles(PeriodFields)').length).toEqual(1);
    });

    it('Adds <FormHelperText/> with error message when an errorText is present', () => {
        const props = {
            ...defaultProps,
            periodType: 'Daily',
            errorText: 'Bad input',
        };
        const wrapper = shallow(<Form {...props} />);
        expect(wrapper.find('Fragment').children().length).toEqual(3);
        expect(wrapper.find('.error.helper').length).toEqual(1);
    });

    it('Renders only a <FormHelperText/> with error message when periodTypes is missing (load error)', () => {
        const props = {
            ...defaultProps,
            periodTypes: null,
            errorText: 'Could not load periodTypes',
        };
        const wrapper = shallow(<Form {...props} />);
        expect(wrapper.find('Fragment').children().length).toEqual(1);
        expect(wrapper.find('.error.helper').length).toEqual(1);
    });

    it('Renders a parsed period below the <Select/> and <PeriodFields/> if there is a valid value', () => {
        const props = {
            ...defaultProps,
            periodType: 'WeeklyWednesday',
            value: '2020WedW12',
        };
        const parsedValue = '2020 W12 March 18 - 24';
        const wrapper = shallow(<Form {...props} />);
        expect(wrapper.find('Fragment').children().length).toEqual(3);
        expect(wrapper.find('.helper').length).toEqual(1);
        expect(wrapper.find('.helper').html()).toContain(parsedValue);
        expect(wrapper).toMatchSnapshot();
    });
});
