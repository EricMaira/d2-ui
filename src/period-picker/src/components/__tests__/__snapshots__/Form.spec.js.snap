// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form/> Renders a parsed period below the <Select/> and <PeriodFields/> if there is a valid value 1`] = `
<Fragment>
  <WithStyles(Select)
    label="Period type"
    name="periodType"
    onChange={[MockFunction]}
    options={
      Array [
        Object {
          "label": "Daily",
          "value": "Daily",
        },
        Object {
          "label": "Weekly",
          "value": "Weekly",
        },
        Object {
          "label": "Weekly Wednesday",
          "value": "WeeklyWednesday",
        },
        Object {
          "label": "Weekly Thursday",
          "value": "WeeklyThursday",
        },
        Object {
          "label": "Weekly Saturday",
          "value": "WeeklySaturday",
        },
        Object {
          "label": "Weekly Sunday",
          "value": "WeeklySunday",
        },
        Object {
          "label": "Bi weekly",
          "value": "BiWeekly",
        },
        Object {
          "label": "Monthly",
          "value": "Monthly",
        },
        Object {
          "label": "Bi-monthly",
          "value": "BiMonthly",
        },
        Object {
          "label": "Quarterly",
          "value": "Quarterly",
        },
        Object {
          "label": "Six monthly",
          "value": "SixMonthly",
        },
        Object {
          "label": "Six monthly starting in April",
          "value": "SixMonthlyApril",
        },
        Object {
          "label": "Six monthly starting in November",
          "value": "SixMonthlyNov",
        },
        Object {
          "label": "Yearly",
          "value": "Yearly",
        },
        Object {
          "label": "Financial year starting in April",
          "value": "FinancialApril",
        },
        Object {
          "label": "Financial year starting in July",
          "value": "FinancialJuly",
        },
        Object {
          "label": "Financial year starting in October",
          "value": "FinancialOct",
        },
        Object {
          "label": "Financial year starting in November",
          "value": "FinancialNov",
        },
      ]
    }
    value="WeeklyWednesday"
  />
  <WithStyles(PeriodFields)
    getValue={[MockFunction]}
    onChange={[MockFunction]}
    periodType="WeeklyWednesday"
  />
  <WithStyles(FormHelperText)
    className="helper"
  >
    2020 W12 March 18 - 24
  </WithStyles(FormHelperText)>
</Fragment>
`;
